File: 剑指offer/59.2.队列的最大值.py
class MaxQueue:
    def __init__(self):
        self.que = []
        self.maxque = []

    def max_value(self) -> int:
        if self.maxque:
            return self.maxque[0]
        else:
            return -1

    def push_back(self, value: int) -> None:
        while self.maxque and self.maxque[-1] < value:
            self.maxque.pop(-1)

        self.maxque.append(value)
        self.que.append(value)

    def pop_front(self) -> int:
        if not self.que:
            return -1
        ans = self.que.pop(0)
        if ans == self.maxque[0]:
            self.maxque.pop(0)
        return ans

File: 剑指offer/59.2.队列的最大值.cpp
#include <deque>
using namespace std;
class MaxQueue {
public:
    deque<int> que;
    deque<int> max_que;

    MaxQueue()
    {
        deque<int> que;
        this->que = que;

        deque<int> max_que;
        this->max_que = max_que;
    }

    int max_value()
    {
        if (not max_que.empty()) {
            return max_que.front();
        } else {
            return -1;
        }
    }

    void push_back(int value)
    {
        while (not max_que.empty() && max_que.back() < value) {
            max_que.pop_back();
        }
        max_que.push_back(value);
        que.push_back(value);
    }

    int pop_front()
    {
        if (que.empty()) {
            return -1;
        }
        int ans = que.front();
        if (ans == max_que.front()) {
            max_que.pop_front();
        }
        que.pop_front();
        return ans;
    }
};
